src/Raven.Client/Server/DatabaseRecord.cs

```diff
@@ -41,21 +43,95 @@ public DatabaseRecord(string databaseName)
 
         public VersioningConfiguration VersioningConfiguration;
 
+        public void AddIndex(IndexDefinition definition)
+        {
+            if (Transformers != null && Transformers.ContainsKey(definition.Name))
+            {
+                throw new IndexOrTransformerAlreadyExistException($"Tried to create an index with a name of {definition.Name}, but a transformer under the same name exists");
+            }
+
+            var lockMode = IndexLockMode.Unlock;
+
+            IndexDefinition existingDefinition;
+            if (Indexes.TryGetValue(definition.Name, out existingDefinition))
+            {
+                if (existingDefinition.LockMode != null)
+                    lockMode = existingDefinition.LockMode.Value;
+
+                var result = existingDefinition.Compare(definition);
+                if (result != IndexDefinitionCompareDifferences.All)
+                {
+                    result &= ~IndexDefinitionCompareDifferences.Etag;
+
+                    if (result == IndexDefinitionCompareDifferences.None)
+                        return;
+                }
+            }
+
+            if (lockMode == IndexLockMode.LockedIgnore)
+                return;
+
+            if (lockMode == IndexLockMode.LockedError)
+                throw new IndexOrTransformerAlreadyExistException($"Cannot edit existing index {definition.Name} with lock mode {lockMode}");
+
+            Indexes[definition.Name] = definition;
+        }
+
+        public void AddIndex(AutoIndexDefinition definition)
+        {
+            if (Transformers != null && Transformers.ContainsKey(definition.Name))
+            {
+                throw new IndexOrTransformerAlreadyExistException($"Tried to create an index with a name of {definition.Name}, but a transformer under the same name exists");
+            }
+
+            var lockMode = IndexLockMode.Unlock;
+
+            AutoIndexDefinition existingDefinition;
+            if (AutoIndexes.TryGetValue(definition.Name, out existingDefinition))
+            {
+                if (existingDefinition.LockMode != null)
+                    lockMode = existingDefinition.LockMode.Value;
+
+                var result = existingDefinition.Compare(definition);
+                result &= ~IndexDefinitionCompareDifferences.Etag;
+
+                if (result == IndexDefinitionCompareDifferences.None)
+                    return;
+
+                result &= ~IndexDefinitionCompareDifferences.LockMode;
+                result &= ~IndexDefinitionCompareDifferences.Priority;
+
+                if (result != IndexDefinitionCompareDifferences.None)
+                    throw new NotSupportedException($"Can not update auto-index: {definition.Name}");
+            }
+
+            if (lockMode == IndexLockMode.LockedIgnore)
```

Comment:
1. LockMode -> now when I'm thinking about this then it is probably not needed
2. Priority - indexing thread priority